### PR TITLE
Enable non-nullable reference types

### PR DESCRIPTION
**Pull request by Mattias1**
## ✨ What's this?
Enable non-nullable reference types for this library.
Projects or libraries depending on this library can now trust that objects marked as non nullable are, in fact, not null. They also receive warnings if they use a nullable object from this library but forget to check on nullability.

### 🔗 Relationships
There is no issue for non-nullable reference types in this project, though there are in other projects, e.g. https://github.com/beardgame/utilities/issues/184.

## 🔍 Why do we want this?
Projects or libraries depending on this library can now trust that objects marked as non nullable are, in fact, not null. They also receive warnings (or errors) if they use a nullable object from this library but forget to check on nullability.
Apart from that, the library itself is also less likely to have null related errors in the future.

## 🏗 How is it done?
Enable the nullable warnings, fix them and then treat the warnings as errors.
There's a separate commit to make things work with visual studio (you guys seem to use Rider). I can remove that commit if you want but IMO it can't hurt to keep the changes for future contributors using visual studio.

### 💥 Breaking changes
- C#7 or lower might no longer be supported. Not sure.
- You can no longer resolve null as a dependency (but you can still resolve the default of a struct).

### 🔬 Why not another way?
- In ListControl.cs there are exceptions thrown if itemSource is null. Now, in most cases they can't be null (because it has nullability checks in the methods calling the one with the exception.
You can also use the null forgiving operator (`!`) there. I went with the exceptions in case you ever call the method from another method and forget to check nullability there. Also, it's more explicit this way.
- I also chose not to introduce any breaking changes due to changes in a public API. So everything that could have null in the past should still accept null now. The reason is to not have any breaking changes ;)
- I might have introduced a 'breaking change' in case an object could receive null in the past, but the library never checked on it once. In that case I've missed it. Maybe you want to look out for these, or you'll receive warnings (or errors) when you use this library.
This is only a breaking change if you treat nullable warnings as errors in the project using this library. Otherwise it's a warning. I think this is a good thing, because not checking for null at all on a nullable field is probably a bug waiting to happen.

### 🦋 Side effects
In some cases if you violate some assumptions of the library you might now see an InvalidOperationException with a description, rather than just a NullReferenceException.

## 💡 Review hints
With one exception this should not have any effect on any logic inside the library or API as exposed to the outside world.
The exception is one small fix with respect to the Control.CanBeFocused property, but that should be a bugfix, not a bug introduction ;)

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,27 @@
+; EditorConfig to support per-solution formatting.
```
Comment (tomrijnbeek) [inline .editorconfig:1]: Adding an .editorconfig sounds like a change that is out of scope for this PR.

Switching to an .editorconfig sounds like a decision/discussion we should have across all Bearded libraries.

File: Bearded.UI/Controls/CompositeControl.cs [hunk 2]
```diff
@@ -46,7 +47,13 @@ public void RemoveAllChildren()
             }
         }
 
-        public bool FocusDescendant(Control control) => Parent.FocusDescendant(control);
+        public bool FocusDescendant(Control control)
+        {
+            if (Parent == null)
```
Comment (tomrijnbeek) [inline Bearded.UI/Controls/CompositeControl.cs:52]: nit: I prefer always using brackets, even for single-line ifs.

File: Bearded.UI/Controls/Control.cs [hunk 3]
```diff
@@ -7,23 +7,35 @@ namespace Bearded.UI.Controls
...
+        }
 
         private bool isVisible = true;
         public bool IsVisible
         {
             get => isVisible;
             set
             {
-                if (isVisible == value) return;
+                if (isVisible == value)
```
Comment (tomrijnbeek) [inline Bearded.UI/Controls/Control.cs:37]: Either use brackets or revert to the previous format.

File: Bearded.UI/Controls/Control.cs [hunk 4]
```diff
@@ -90,21 +119,17 @@ public void SetFrameNeedsUpdateIfNeeded()
 
         public virtual void SetFrameNeedsUpdate()
         {
-            frameNeedsUpdate = true;
-        }
+            if (Parent == null)
```
Comment (tomrijnbeek) [inline Bearded.UI/Controls/Control.cs:122]: I think this will cause bugs, since you may call this before adding the control to the hierarchy. In this case, that means it's no harm done: the frame should never be recalculated if the parent is null (because if the parent is null, we'll throw an exception anyway).

File: Bearded.UI/Controls/Control.cs [hunk 5]
```diff
@@ -64,6 +90,9 @@ public virtual bool TryFocus()
             if (IsFocused)
                 return true;
 
+            if (Parent == null)
```
Comment (tomrijnbeek) [inline Bearded.UI/Controls/Control.cs:93]: nit: should we check this before the `IsFocused` check?

File: Bearded.UI/Controls/Control.cs [hunk 6]
```diff
@@ -33,15 +45,29 @@ public bool IsVisible
...
         public bool IsFocused { get; private set; }
-        public bool CanBeFocused { get; protected set; }
+
+        private bool canBeFocused = false;
+        public bool CanBeFocused
+        {
+            get => canBeFocused;
+            protected set
+            {
+                if (!value && IsFocused)
```
Comment (tomrijnbeek) [inline Bearded.UI/Controls/Control.cs:64]: Is this a behavioural change? If so, please address in a separate PR and explain reasoning why the old method was broken.

File: Bearded.UI/NullableAttributes.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+namespace Bearded.UI
+{
+    // To deal with special cases with non nullable reference types, the C# compiler team has made some attributes,
+    // which they then only added for .NET Core. So I've had to add my own copy here.
```
Comment (tomrijnbeek) [inline Bearded.UI/NullableAttributes.cs:6]: This is fair enough.

This should be fixed when we migrate this library to OpenTK 4 stable.

Suggestion: mark this as Obsolete to make sure we don't forget to replace this with the .NET Core one. Feel free to file an issue too.

**Review by tomrijnbeek — CHANGES_REQUESTED** [review]
Thanks.

These are generally some really good changes. It seems like most of the changes are very null-strict instead of null-forgiving, so we'll have to find out if this leads to bugs (or reveals bugs, perhaps), but we can iterate on that as needed.

The majority of the comments are styling-related.

File: .editorconfig (hunk 1 above)
Comment (Mattias1) [reply .editorconfig:1]: Without it the auto-formatting doesn't work in visual studio, but I'll remove it.

File: Bearded.UI/Controls/CompositeControl.cs (hunk 2 above)
Comment (Mattias1) [reply Bearded.UI/Controls/CompositeControl.cs:52]: I agree, but not using brackets is consistent with the styling used in the rest of the repo.

I can use brackets wherever I've added code, but that would be almost the only places (for single lines).

File: Bearded.UI/Controls/Control.cs (hunk 4 above)
Comment (Mattias1) [reply Bearded.UI/Controls/Control.cs:122]: Do you want me to revert this one? Theoretically one could call this, be lucky and set the parent before the Frame getter is called and not crash otherwise, so this is a logic change.

File: Bearded.UI/Controls/Control.cs (hunk 6 above)
Comment (Mattias1) [reply Bearded.UI/Controls/Control.cs:64]: Will revert and file an issue.

**Comment by Mattias1** [comment]
The build fails (probably because xUnit doesn't support netstandard but it is a target in the test project).
Can I ignore the build fail? Do I need to fix that in this PR?

**Comment by tomrijnbeek** [comment]
> The build fails (probably because xUnit doesn't support netstandard but it is a target in the test project).
> Can I ignore the build fail? Do I need to fix that in this PR?

I think you may need to add a NuGet dependency on `Microsoft.NET.Test.Sdk` for the project.

It would be good not to push a broken build to master, because then we won't be releasing any new versions to NuGet.

File: .editorconfig (hunk 1 above)
Comment (tomrijnbeek) [reply .editorconfig:1]: It's a fair point, and something we should do a better job of.

Filed beardgame/utilities#212 to track this across all Bearded libraries.

File: Bearded.UI/Controls/CompositeControl.cs (hunk 2 above)
Comment (tomrijnbeek) [reply Bearded.UI/Controls/CompositeControl.cs:52]: If this is consistent with the rest of the code, it's fine to keep. We should then probably do a mass change to make this more consistent.

File: Bearded.UI/Controls/Control.cs (hunk 4 above)
Comment (tomrijnbeek) [reply Bearded.UI/Controls/Control.cs:122]: Yes, let's continue to allow this. We can reconsider this logic change if this would reveal a bug.

File: Bearded.UI/Controls/Control.cs [hunk 8]
```diff
@@ -116,12 +141,18 @@ private void recalculateFrame()
...
+                throw new InvalidOperationException("This control doesn't have a parent.");
+
+            Parent.Remove(this);
+        }
 
         internal void AddTo(IControlParent parent)
         {
             if (Parent != null)
-                throw new InvalidOperationException();
+                throw new InvalidOperationException("This control already has a parent.");
```
Comment (amulware) [inline Bearded.UI/Controls/Control.cs:155]: Do we actually need this check on this internal method which we have full control over? I'd rather make that a Debug.Assert perhaps?

Also ping @tomrijnbeek for how we want to handle cases like this in general.

File: Bearded.UI/Controls/Control.cs [hunk 9]
```diff
@@ -131,7 +162,7 @@ internal void AddTo(IControlParent parent)
         internal void RemoveFrom(IControlParent parent)
         {
             if (parent != Parent)
-                throw new InvalidOperationException();
+                throw new InvalidOperationException("This control has a different parent.");
```
Comment (amulware) [inline Bearded.UI/Controls/Control.cs:165]: Similar to the above, maybe this should be a Debug.Assert?

File: Bearded.UI/Controls/implementations/ListControl.cs [hunk 10]
```diff
@@ -229,6 +235,9 @@ public void Reload()
 
         private void calculateTotalHeight()
         {
+            if (itemSource == null)
+                throw new InvalidOperationException("The ItemSource is not set.");
```
Comment (amulware) [inline Bearded.UI/Controls/implementations/ListControl.cs:239]: This method is only called from a place where we already verified this isn't null. Do we have to check it again?

File: Bearded.UI/Controls/implementations/ListControl.cs [hunk 11]
```diff
@@ -265,6 +274,9 @@ private void updateContentLimits()
 
         private void removeCellsUpwards()
         {
+            if (itemSource == null)
+                throw new InvalidOperationException("The ItemSource is not set.");
```
Comment (amulware) [inline Bearded.UI/Controls/implementations/ListControl.cs:278]: I think my comments are getting repetitive and boil down to this: Should we throw in private methods like this at all? It feels smelly to me. We should write our code so that it won't call private methods in invalid circumstances, instead of checking everywhere I think. (I'll stop adding the same comment to all the below cases.)

File: Bearded.UI/Controls/implementations/ListControl.cs [hunk 12]
```diff
@@ -327,35 +342,41 @@ private void addCellsDownwards()
...
+
             var height = itemSource.HeightOfItemAt(index);
             var top = bottom - height;
 
             var cell = createCellIfVisible(index, bottom, top, height);
 
             if (cell.Control != null)
-                cells.AddFirst(cell);
+                cells.AddFirst((cell.Control, cell.Index, cell.Offset, cell.Height));
```
Comment (amulware) [inline Bearded.UI/Controls/implementations/ListControl.cs:374]: This looks like we should be able to write it nicer instead of having to recreate almost the same struct.

File: Bearded.UI/Controls/implementations/ListControl.cs (hunk 10 above)
Comment (Mattias1) [reply Bearded.UI/Controls/implementations/ListControl.cs:239]: Answering your question here, as on a private method and internal method one might have different opinions, though I think an internal method should have the same checks as a public method should, but that's my opinion.


The C# compiler does not know that itemSource is or isn't null here, so it assumes it can be null, and thus gives an error (or warning). I don't know if the compiler doesn't know this on purpose, or because it can't evaluate that deeply.

You can solve this error by either checking for null and then throwing an exception, or by using the null-forgiving operator (`!`). I don't know if a Debug.Assert works, but I think not.
The pro of using the `!` is that it doesn't make your code ugly by these two throwing lines (and that your code is a tiiiny bit faster, for it doesn't have to evaluate an if that always has the same outcome).
The con is that if you ever call this method from another place, and you forget to check for null, the compiler will not warn you about it. And the fact that you have nullable errors turned on, and the compiler not warning you about it may give you a false sense of security causing you to forget it.

The other alternative I guess is to not treat the nullable warnings as errors, and just leave the warning there and ignore it. Personally I think that teaching yourself to ignore warnings is a really bad choice though.


I don't mind using the `!` here, it's your project after all, but then it's on you (or Tom I suppose) to be careful even when your compiler tells you all is well.

File: Bearded.UI/Controls/implementations/ListControl.cs (hunk 12 above)
Comment (Mattias1) [reply Bearded.UI/Controls/implementations/ListControl.cs:374]: Do you know how? These structs are different (the first entry has a different type - nullable and non nullable).
I really don't want to change the type of `cells` to always assume the control can be null.

I'm assuming here that the compiler optimizes this to the same code, but I didn't check that.

File: Bearded.UI/Controls/implementations/ListControl.cs (hunk 11 above)
Comment (Mattias1) [reply Bearded.UI/Controls/implementations/ListControl.cs:278]: :) I'll fix a problem of a comment in the entire file. I might miss it if the same thing happends in a different file, so a reminder there is nice to have, but yeah, I'll change it here as well (if we want to change this - see my answer on the other comment).

File: Bearded.UI/Controls/implementations/ListControl.cs (hunk 10 above)
Comment (amulware) [reply Bearded.UI/Controls/implementations/ListControl.cs:239]: How about passing in the verified-non-null collection into the private methods?

That way we get compile time safety (and I fully agree, we should not ignore these warnings, and we should in fact make them errors!), and we don't have to check all over the place.

File: Bearded.UI/Controls/implementations/ListControl.cs (hunk 12 above)
Comment (amulware) [reply Bearded.UI/Controls/implementations/ListControl.cs:374]: I was hoping we could make the whole tuple nullable but looks like we need the associated data in either case. It's a little annoying, but I don't see an easier way of improving it. It could probably be refactored, but perhaps that's for another time.

File: Bearded.UI/NullableAttributes.cs (hunk 7 above)
Comment (Mattias1) [reply Bearded.UI/NullableAttributes.cs:6]: Issue created: https://github.com/beardgame/ui/issues/20

File: Bearded.UI/Controls/Control.cs (hunk 6 above)
Comment (Mattias1) [reply Bearded.UI/Controls/Control.cs:64]: The issue: https://github.com/beardgame/ui/issues/21

File: Bearded.UI/Controls/Control.cs (hunk 8 above)
Comment (Mattias1) [reply Bearded.UI/Controls/Control.cs:155]: Oh haha, the check is already there, I only added a message to make it more clear why it's throwing. I won't revert these ones if you don't mind.

File: Bearded.UI/Controls/implementations/ListControl.cs (hunk 10 above)
Comment (Mattias1) [reply Bearded.UI/Controls/implementations/ListControl.cs:239]: I applied the passing it as parameter everywhere, but don't like it, because it has a pertty big fallout radius.

It did spot two places where there wasn't checked for nullability yet though, so I guess it's better this way. I'll commit it and let you decide if you can allow the ugliness of passing it in almost every method.

**Comment by Mattias1** [comment]
> 
> 
> > The build fails (probably because xUnit doesn't support netstandard but it is a target in the test project).
> > Can I ignore the build fail? Do I need to fix that in this PR?
> 
> I think you may need to add a NuGet dependency on `Microsoft.NET.Test.Sdk` for the project.
> 
> It would be good not to push a broken build to master, because then we won't be releasing any new versions to NuGet.

I already added that package, which fixes the tests for the net462 target, but not for netstandard. Do you really need to target netstandard? Given https://xunit.net/docs/why-no-netstandard I doubt I can fix the build easily, and it feels outside the scope of this PR. I think the build already fails if you add an empty PR right now.

TODO: Let me try that actually.

**Comment by tomrijnbeek** [comment]
> I already added that package, which fixes the tests for the net462 target, but not for netstandard. Do you really need to target netstandard? Given https://xunit.net/docs/why-no-netstandard I doubt I can fix the build easily, and it feels outside the scope of this PR. I think the build already fails if you add an empty PR right now.

I doubt it just started breaking out of nowhere, so an empty PR should succeed just fine.

Since we're switching to .NET Core as part of a migration to OpenTK 4, I am fine if you just want to remove the netstandard targeting for now, since the targeting is going to change completely anyway. That being said, I've had xunit run fine on netstandard builds before, so it's a bit baffling this just started breaking out of nowhere.

**Review by tomrijnbeek — COMMENTED** [review]
Looks like there's still some outstanding comments that need to be resolved, right? Otherwise, I think we're moving in the right direction here.

**Comment by Mattias1** [comment]
> 
> 
> Looks like there's still some outstanding comments that need to be resolved, right? Otherwise, I think we're moving in the right direction here.

amulware should see if the things he commented on are ok now. I think all other comments are adressed now.

File: Bearded.UI/Navigation/DependencyResolver.cs [hunk 13]
```diff
@@ -17,7 +20,19 @@ public T Resolve<T>()
...
-        public bool TryResolve<T>(out T value)
+        public bool TryResolve<T>([NotNullWhen(returnValue: true)] out T? value) where T : class
+        {
+            if (dict.TryGetValue(typeof(T), out var obj))
+            {
+                value = (T) obj;
+                return true;
+            }
+
+            value = null;
```
Comment (tomrijnbeek) [inline Bearded.UI/Navigation/DependencyResolver.cs:31]: nit: I find `value = default` a bit nicer

**Review by tomrijnbeek — APPROVED** [review]
I will leave it to Paul to say whether he likes the list control changes. I don't feel strongly either way.

File: Bearded.UI/Navigation/DependencyResolver.cs (hunk 13 above)
Comment (amulware) [reply Bearded.UI/Navigation/DependencyResolver.cs:31]: I think that's a different semantic meaning, and being explicit that we're returning null here is fine and even good.

File: Bearded.UI/Controls/implementations/ListControl.cs [hunk 14]
```diff
@@ -95,11 +95,11 @@ public override void MouseScrolled(MouseScrollEventArgs eventArgs)
 
             if (offsetAfter > offsetBefore)
             {
-                onScrollDown();
+                onScrollDown(itemSource);
```
Comment (amulware) [inline Bearded.UI/Controls/implementations/ListControl.cs:98]: Ok, I see what you mean now... there's not many checks, but the collection is passed around a whole lot indeed. I'm happy merging it like this, but perhaps for the future, I'd suggest we try refactoring all the private methods into a separate class wrapping the ItemSource internally, that has all the 'nullable behaviour' on it, so there's no passing and checks only at the entry points to 'private behaviour'. Though perhaps that'll end up with having to pull other fields into that class that we need in other places, and there's just no great solution here.

**Review by amulware — APPROVED** [review]

File: Bearded.UI/Controls/implementations/ListControl.cs [hunk 15]
```diff
@@ -110,29 +110,35 @@ public override void MouseScrolled(MouseScrollEventArgs eventArgs)
...
+        private void onScrollDown(IListItemSource itemSource)
         {
-            addCellsDownwards();
-            removeCellsDownwards();
+            addCellsDownwards(itemSource);
+            removeCellsDownwards(itemSource);
         }
         
-        public void OnAppendItems(int addedCount)
+        public void OnAppendItems(int addedCount, IListItemSource itemSource)
```
Comment (amulware) [inline Bearded.UI/Controls/implementations/ListControl.cs:141]: This method is public, I'm pretty sure it should not have this new parameter.

**Review by amulware — CHANGES_REQUESTED** [review]
Missed something that (luckily) got flagged as a merge conflict.

**Comment by amulware** [comment]
Was meaning to merge the PR, hence the commit, though I'm not sure I got it right - the build is failing.

**Comment by amulware** [comment]
From the build:
```
error CS8618: Non-nullable field 'contentContainer' is uninitialized. Consider declaring the field as nullable. 
```

Let's address this too, should be easy. :)

**Review by tomrijnbeek — APPROVED** [review]

**Review by amulware — APPROVED** [review]
